File: backend/F1Fantasy/Controllers/ConstructorStandingController.cs
using F1Fantasy.Models;
using F1Fantasy.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace F1Fantasy.Controllers;

[Authorize]
[ApiController]
[Route("api/[controller]")]
public class ConstructorStandingController : ControllerBase
{
    private readonly ConstructorStandingService _service;
    private readonly ILogger<ConstructorStandingController> _logger;

    public ConstructorStandingController(ConstructorStandingService service, ILogger<ConstructorStandingController> logger)
    {
        _service = service;
        _logger = logger;
    }

    [HttpGet("season/{season}")]
    public async Task<ActionResult<ConstructorStandingsList>> GetConstructorStandingsBySeason(string season)
    {
        try
        {
            _logger.LogInformation("Request received: Get constructor standings for season {Season}", season);
            var standings = await _service.GetConstructorStandingsBySeasonAsync(season);
            
            if (standings == null)
            {
                _logger.LogWarning("No constructor standings found for season {Season}", season);
                return NotFound(new ErrorResponse { Message = $"No constructor standings found for season {season}" });
            }

            return Ok(standings);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing request for constructor standings season {Season}", season);
            return StatusCode(500, new ErrorResponse { Message = "An error occurred while processing your request" });
        }
    }

    [HttpGet("season/{season}/round/{round}")]
    public async Task<ActionResult<ConstructorStandingsList>> GetConstructorStandingsByRound(string season, string round)
    {
        try
        {
            _logger.LogInformation("Request received: Get constructor standings for season {Season} round {Round}", season, round);
            var standings = await _service.GetConstructorStandingsByRoundAsync(season, round);
            
            if (standings == null)
            {
                _logger.LogWarning("No constructor standings found for season {Season} round {Round}", season, round);
                return NotFound(new ErrorResponse { Message = $"No constructor standings found for season {season} round {round}" });
            }

            return Ok(standings);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing request for constructor standings season {Season} round {Round}", season, round);
            return StatusCode(500, new ErrorResponse { Message = "An error occurred while processing your request" });
        }
    }

    [HttpGet("season/{season}/round/{round}/constructor/{constructorId}")]
    public async Task<ActionResult<ConstructorStandingEntry>> GetConstructorStandingByConstructor(string season, string round, string constructorId)
    {
        try
        {
            _logger.LogInformation("Request received: Get constructor standing for {ConstructorId} season {Season} round {Round}", 
                constructorId, season, round);
            var standing = await _service.GetConstructorStandingByConstructorAsync(season, round, constructorId);
            
            if (standing == null)
            {
                _logger.LogWarning("No constructor standing found for {ConstructorId} season {Season} round {Round}", 
                    constructorId, season, round);
                return NotFound(new ErrorResponse { Message = $"No constructor standing found for {constructorId} in season {season} round {round}" });
            }

            return Ok(standing);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing request for constructor standing {ConstructorId} season {Season} round {Round}", 
                constructorId, season, round);
            return StatusCode(500, new ErrorResponse { Message = "An error occurred while processing your request" });
        }
    }

    [HttpGet("cached")]
    public async Task<ActionResult<List<ConstructorStandingsList>>> GetCachedStandings()
    {
        try
        {
            _logger.LogInformation("Request received: Get all cached constructor standings");
            var standings = await _service.GetCachedStandingsAsync();
            return Ok(standings);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing request for cached constructor standings");
            return StatusCode(500, new ErrorResponse { Message = "An error occurred while processing your request" });
        }
    }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/controller/ConstructorStandingController.java
package no.f1fantasy.controller;

import java.util.HashMap;
import java.util.Map;

import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import no.f1fantasy.entity.ConstructorStanding;
import no.f1fantasy.service.ConstructorStandingService;

@RestController
@RequestMapping("/api/constructor-standing")
@PreAuthorize("isAuthenticated()")
@SuppressWarnings("null")
public class ConstructorStandingController {

    private final ConstructorStandingService constructorStandingService;

    public ConstructorStandingController(ConstructorStandingService constructorStandingService) {
        this.constructorStandingService = constructorStandingService;
    }

    @GetMapping("/season/{season}")
    public ResponseEntity<?> getConstructorStandingsBySeason(@PathVariable String season) {
        try {
            java.util.List<ConstructorStanding> standings = constructorStandingService.getConstructorStandingsBySeason(season);
            if (standings == null || standings.isEmpty()) {
                return ResponseEntity.status(404).body(errorMap("No constructor standings found for season " + season));
            }
            return ResponseEntity.ok(standings);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching constructor standings for season " + season));
        }
    }

    @GetMapping("/season/{season}/round/{round}")
    public ResponseEntity<?> getConstructorStandingsByRound(@PathVariable String season, @PathVariable String round) {
        try {
            java.util.List<ConstructorStanding> standings = constructorStandingService.getConstructorStandingsByRound(season, round);
            if (standings == null || standings.isEmpty()) {
                return ResponseEntity.status(404).body(errorMap("No constructor standings found for season " + season + ", round " + round));
            }
            return ResponseEntity.ok(standings);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching constructor standings for season " + season + ", round " + round));
        }
    }

    @GetMapping("/season/{season}/round/{round}/constructor/{constructorId}")
    public ResponseEntity<?> getConstructorStandingByConstructor(@PathVariable String season, @PathVariable String round, @PathVariable String constructorId) {
        try {
            java.util.Optional<ConstructorStanding> standing = constructorStandingService.getConstructorStandingByConstructor(season, round, constructorId);
            if (!standing.isPresent()) {
                return ResponseEntity.status(404).body(errorMap("No constructor standing found for constructor " + constructorId + " in season " + season + ", round " + round));
            }
            return ResponseEntity.ok(standing.get());
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching constructor standing for constructor " + constructorId));
        }
    }

    private Map<String, String> errorMap(String message) {
        Map<String, String> error = new HashMap<>();
        error.put("error", message);
        return error;
    }
}
